### PR TITLE
Implement @phpstan-allow-private-mutation

### PR DESCRIPTION
**Pull request by zonuexe**
resolve https://github.com/phpstan/phpstan/issues/7778, refs https://github.com/phpstan/phpstan-src/pull/1609

### CODE REVIEW

**Comment by zonuexe** [comment]
`ReadOnlyByPhpDocPropertyAssignRuleTest` alone will pass with e2d9db8fc29efb52a847de6df704b7a540c0331e patch, but you probably need to rethink how `ResolvedPhpDocBlock::$phpDocNode` is initialized.

File: src/PhpDoc/ResolvedPhpDocBlock.php [hunk 1]
```diff
@@ -192,7 +198,9 @@ public function merge(array $parents, array $parentPhpDocBlocks): self
 		// skip $result->nameScope
 		$result->templateTypeMap = $this->templateTypeMap;
 		$result->templateTags = $this->templateTags;
-		// skip $result->phpDocNodeResolver
+		if (isset($this->phpDocNodeResolver)) {
+			$result->phpDocNodeResolver = $this->phpDocNodeResolver;
+		}
```
Comment (zonuexe) [inline src/PhpDoc/ResolvedPhpDocBlock.php]:
```
Error: Property PHPStan\PhpDoc\ResolvedPhpDocBlock::$phpDocNode (PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocNode) in isset() is not nullable.
Error: Property PHPStan\PhpDoc\ResolvedPhpDocBlock::$phpDocNodeResolver (PHPStan\PhpDoc\PhpDocNodeResolver) in isset() is not nullable.
 ------ ------------------------------------------------------------------ 
  Line   src/PhpDoc/ResolvedPhpDocBlock.php                                
 ------ ------------------------------------------------------------------ 
  225    Property PHPStan\PhpDoc\ResolvedPhpDocBlock::$phpDocNode          
         (PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocNode) in isset() is not    
         nullable.                                                         
  [24](https://github.com/phpstan/phpstan-src/actions/runs/3044344145/jobs/4904629542#step:11:25)3    Property PHPStan\PhpDoc\ResolvedPhpDocBlock::$phpDocNodeResolver  
         (PHPStan\PhpDoc\PhpDocNodeResolver) in isset() is not nullable.   
 ------ ------------------------------------------------------------------
```

https://github.com/phpstan/phpstan-src/actions/runs/3044344145/jobs/4904629542#step:11:9

**Comment by zonuexe** [comment]
@ondrejmirtes Could you please review this PR?

File: src/PhpDoc/ResolvedPhpDocBlock.php [hunk 2]
```diff
@@ -192,7 +198,9 @@ public function merge(array $parents, array $parentPhpDocBlocks): self
 		// skip $result->nameScope
 		$result->templateTypeMap = $this->templateTypeMap;
 		$result->templateTags = $this->templateTags;
-		// skip $result->phpDocNodeResolver
+		if (isset($this->phpDocNodeResolver)) {
+			$result->phpDocNodeResolver = $this->phpDocNodeResolver;
```
Comment (ondrejmirtes) [inline src/PhpDoc/ResolvedPhpDocBlock.php]: Hi, you¨re adding a simple boolean value which there are already many present. These changes to phpDocNode and phpDocNodeResolver are not required to make this work, and cause these build failures.

If the PR doesn't work without these changes, please observe how different boolean values are handled here, it shouldn't be hard to make them work.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

**Comment by zonuexe** [comment]
Sorry, I accidentally submitted a re-review. Ignore it for once.

**Comment by ondrejmirtes** [comment]
I've updated the PR with the correct fix.

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you.

**Pull request merged**